File: Hello Algorithm/Hello Algorithm_Swift/Hello Algorithm/Hello Algorithm/链表/CycleLinkedList.swift
//
//  CycleLinkedList.swift
//  Hello Algorithm
//
//  Created by mac on 2021/8/20.
//

import Foundation

public let CYCLE_LINKED_LIST_NOT_FONUD = -1

fileprivate class CycleLinkedListNode <T: Equatable> : CustomStringConvertible {
    var next: CycleLinkedListNode?
    var element: T
    var prev: CycleLinkedListNode?
    init(next: CycleLinkedListNode?,element: T,prev: CycleLinkedListNode?) {
        self.prev = prev
        self.element = element
        self.next = next;
    }
    
    var description: String {
        var des = ""
        if let thePrev = prev {
            des = des.appending("\(thePrev.element)")
        } else {
            des = des.appending("nil")
        }
        
        des = des.appending("_")
        des = des.appending("【\(element)】");
        des.append("_")
        
        if let theNext = next {
            des = des.appending("\(theNext.element)")
        } else {
            des = des.appending("nil")
        }
        
        return des
    }
}

class CycleLinkedList <T: Equatable> : CustomStringConvertible {
    private var _size = 0 //链表节点的数量
    private var _first: CycleLinkedListNode<T>? //链表的头结点
    private var _last: CycleLinkedListNode<T>? //链表的尾结点
    
    //MARK: - 查
    
    /// 链表的节点数量
    /// - Returns: description
    func size() -> Int {
        _size
    }
    
    /// 链表是否为空
    /// - Returns: description
    func isEmpty() -> Bool {
        _size == 0
    }
    
    
    /// 链表中是否含有element
    /// - Parameter element: element
    /// - Returns: result
    func contains(element: T) -> Bool {
        !(indexOf(element: element) == CYCLE_LINKED_LIST_NOT_FONUD)
    }
    
    
    /// 获取元素element的索引
    /// - Parameter element: element
    /// - Returns: index
    func indexOf(element: T) -> Int {
        if isEmpty() {
            return CYCLE_LINKED_LIST_NOT_FONUD
        }
        
        var node = _first
        for i in 0..._size {
            if let theNode = node {
                if theNode.element == element {
                    return i
                }
            }
            node = node?.next
        }
        
        return CYCLE_LINKED_LIST_NOT_FONUD
    }
    
    
    /// 获取指定索引的元素
    /// - Parameter index: index
    /// - Returns: element
    func get(index: Int) -> T {
        nodeAt(index: index).element
    }
    
    //MARK: - 增加
    
    /// 向链表中指定索引处增加元素
    /// - Parameters:
    ///   - element: element
    ///   - index: index
    /// - Returns: void
    func add(element: T,index: Int) -> Void {
        /*
         对于一般情况（非头节点、尾结点）可以使用如下代码
         let node = nodeAt(index: index) //获取index位置的节点
         let prev = node.prev // node的前置节点
         //创建新的节点，并试 newNode.prev = node.prev, newNode.next = node
         let newNode = CycleLinkedListNode(next: node, element: element, prev: prev)
         //令node.prev = newNode
         node.prev = newNode
        */
        
        //插入的位置是最后
        if index == _size {
            let oldLast = _last
            let newNode = CycleLinkedListNode(next: nil, element: element, prev: _last)
            _last = newNode
            // 考虑原链表 isEmpty的情况，oldLast不为空，则链表不为空
            if let realOldLast = oldLast {
                realOldLast.next = newNode
            }
            //oldLast为空，原链表为空
            else {
                _first = newNode
            }
        } else {
            let node = nodeAt(index: index)
            let newNode = CycleLinkedListNode(next: node, element: element, prev: node.prev)
            let oPrev = node.prev
            node.prev = newNode
            //如果prv不为空，则说明插入的节点不是头结点
            if let prev = oPrev {
                prev.next = newNode
            }
            //如果prev为空,则说明插入的节点是头结点
            else {
                _first = newNode
            }
        }
        _size = _size + 1
    }
    
    /// 向链表最后的位置增加节点
    /// - Parameter element: element
    /// - Returns: void
    func add(element: T) -> Void {
        add(element: element, index: _size)
    }
    
    // MARK: - 删除
    
    /// 移除index索引的元素
    /// - Parameter index: index
    /// - Returns: description
    func remove(index: Int) -> T {
        let node = nodeAt(index: index)
        let elementToRemove = node.element
        let oPrev = node.prev
        let oNext = node.next
        oPrev?.next = oNext
        oNext?.prev = oPrev
        if oPrev == nil { //prev节点为空,则说明移除的是头结点
            _first = oNext
        }
        if oNext == nil { //next节点为空，则说明移除的是尾结点
            _last = oPrev
        }
        _size = _size - 1
        return elementToRemove
    }
    
    func clear() -> Void {
        _first = nil
        _last = nil
        _size = 0
    }
    
    //MARK: - 改
    
    /// 修改index索引的元素
    /// - Parameters:
    ///   - index: index
    ///   - element: element
    /// - Returns: old element
    func set(index:Int, element: T) -> T {
        let node = nodeAt(index: index)
        let oldElement = node.element
        node.element = element
        return oldElement
    }
    
    //MARK: - overwrite
    var description: String {
        var des = "Size = \(_size),["
        var oNode = _first
        for _ in 0..<_size {
            if let node = oNode {
                des = des.appending(node.description)
                des = des.appending(",")
                oNode = node.next
            }
        }
        des = des.appending("]")
        return des
    }
    
    //MARK: - private
    
    ///  获取指定索引的node
    /// - Parameter index: index
    /// - Returns: node
    private func nodeAt(index: Int) -> CycleLinkedListNode<T> {
        //检查index的合法性
        checkIndex(index: index)
        var node: CycleLinkedListNode<T>
        
        // 如果index小于size的一半，则正向遍历
        if index < (_size >> 1) {
            node = _first!
            for _ in 0..<index {
                node = node.next!
            }
        } else {
            node = _last!
            for _ in (index..<(_size - 1)).reversed() {
                node = node.prev!
            }
        }
        return node
    }
    
    
    /// 检查索引是否合法
    /// - Parameter index: index
    /// - Returns: void
    private func checkIndex(index: Int) -> Void {
        if index < 0 || index > _size - 1 {
            fatalError("index:\(index) is out of bounds(size = \(_size))")
        }
    }
    
}

File: Hello Algorithm/Hello Algorithm_Java/src/com/algorithm/List/CycleLinkedList.java
package com.algorithm.List;

import java.lang.ref.SoftReference;

/**
 * 双向循环链表
 */
public class CycleLinkedList<E> {
    // 链表节点的数量
    private int size;
    //首节点
    private Node<E> first;
    //尾结点
    private Node<E> last;

    //没有找到的标识
    public static final int ELEMENT_NOT_FOUND = -1;

    /**
     * 获取链表节点数量
     * @return
     */
    public int size() {
        return size;
    }

    /**
     * 判断链表是否为空
     * @return
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * 判断链表是否包含数据 element
     * @param element
     * @return
     */
    public boolean contains(E element) {
       return indexOf(element) != ELEMENT_NOT_FOUND;
    }

    /**
     * 获取数据 element的索引
     * @param element
     * @return
     */
    public int indexOf(E element) {
        Node<E> cur = first;
        for (int i = 0; i < size; i++) {
            if (element == null) {
                if (cur.element == null) {
                    return i;
                }
            } else {
                if (cur.element.equals(element)) {
                    return i;
                }
            }
            cur = cur.next;
        }
        return ELEMENT_NOT_FOUND;
    }

    /**
     * 获取指定索引的Node节点
     * @param index
     * @return
     */
    public Node<E> get(int index) {
        return node(index);
    }

    /**
     * 向链表末尾添加数据 element
     * @param element
     */
    public void add(E element) {
        add(element,size);
    }

    /**
     * 向指定索引处添加数据 element
     * @param element
     * @param index
     */
    public void add(E element,int index) {
        if (index == size) { //添加链表最后
            Node<E> oldLast = last;
            last = new Node<>(oldLast,element,null);
            if (oldLast == null) {//链表是空的
                //相当于整个链表只有一个节点,first和last都指向这个节点
                first = last;
            } else {//链表不是空的
                //将之前的last.next -> 新的last
                oldLast.next = last;
            }
        } else {//添加到链表其他位置
            Node<E> node = node(index);
            Node<E> newNode = new Node<>(node.prev,element,node);
            node.prev = newNode;
            //判断是否插在头结点
            if (newNode.prev != null) {
                newNode.prev.next = newNode;
            } else { // index == 0
                first = newNode;
            }
        }
        size++;
    }

    /**
     * 移除指定索引的节点
     * @param index
     */
    public E remove(int index) {
        Node<E> node = node(index);
        /* 对于头结点和尾结点以及普通节点分别判断
        if (node.prev == null) { // index == 0
            first = node.next;
            node.next.prev = null;
        } else if (node.next == null) { //移除尾结点
            node.prev.next = node.next;
            last = node.prev;
        } else {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }
        */


        /**
         * 对于一般的情况（不考虑头结点和尾结点）：
         * Node<E> next = node.next;
         * Node<E> prev = node.prev;
         * next.prev = prev;
         * prev.next = next;
         * 但是如果考虑头结点和尾结点，next和prev就有可能为null
         * 所以要分别考虑
         */

        Node<E> next = node.next;
        Node<E> prev = node.prev;

        //只有一个节点的情况也是可以满足的

        if (next == null) {//要移除的节点是尾结点
            last = prev;
        } else {
            next.prev = prev;
        }

        if (prev == null) { //要移除的节点是头结点
            first = next;
        } else {
            prev.next = next;
        }

        size--;
        return node.element;
    }

    /**
     * 清空链表
     */
    public void clear() {
        first = null;
        last = null;
        size = 0;
    }

    /**
     * 设置指定索引节点的数据
     * @param index
     * @param element
     * @return
     */
    public E set(int index,E element) {
        Node<E> node = node(index);
        E oldElement = node.element;
        node.element = element;
        return oldElement;
    }

    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer();
        sb.append("size = " + size + ",[");
        Node<E> node = first;
        for (int i = 0; i < size; i++) {
            sb.append(node).append(",");
            node = node.next;
        }
        sb.append("]");
        return sb.toString();
    }

    // private method
    private Node<E> node(int index) {
        checkIndex(index);
        Node<E> node = null;
        //如果靠近头结点，正向遍历
        if (index < (size >> 1)) {
            node = first;
            for (int i = 0; i < index; i++) {
                node = node.next;
            }
        } else  { //如果比较靠近尾结点则反向遍历
            node = last;
            for (int i = size - 1; i > index; i--) {
                node = node.prev;
            }
        }
        return node;
    }

    private void checkIndex(int index) {
        if (index < 0 || index > size - 1) {
            throw new IndexOutOfBoundsException("index: " + index + "越界,Size=" + size);
        }
    }



    private static class Node<E> {
        // 保存的数据
        E element;
        // 后一个指针
        Node<E> next;
        // 前一个节点
        Node<E> prev;

        Node(Node<E> prev,E element, Node<E> next) {
            this.element = element;
            this.next = next;
            this.prev = prev;
        }

        @Override
        public String toString() {
            StringBuffer sb = new StringBuffer();
            if (prev != null) {
                sb.append(prev.element);
            } else {
                sb.append("null");
            }
            sb.append("_").append(element).append("_");
            if (next != null) {
                sb.append(next.element);
            } else {
                sb.append("null");
            }
            return sb.toString();
        }
    }

}
